main/acle.md

```diff
@@ -2682,6 +2683,10 @@ The following attributes trigger the multi version code generation:
 * If only the `default` version exist it should be linked directly.
 * FMV may be disabled in compile time by a compiler flag. In this
   case the `default` version shall be used.
+* Scope for a group of multiversioned functions is the scope of the default
```

Comment:
I completely agree we want to be able to compile in absence of default declaration. I wrote this to exclude the calling or referring to the dispatched function without the default in scope (rather than the declaration).

My model for this is that the default function _is_ the function, and the extra versions amend that function with different versions for different architecture.
 
This way, if you were to remove all the other version implementations and declarations the code base would still compile.

I think a case like https://godbolt.org/z/qW7hx58bf is emblematic of this point. In this TU there is no indication there is a default version, I think this should require a default version to be declared.

Even in the case you cited (https://github.com/llvm/llvm-project/pull/84405#discussion_r1525583647) they do include the declaration of the default version in all TU's, I believe because it is the idiomatic way to do it?

Reply:
I acknowledge that I'm very junior in my career and that I am may be missing significant nuance in this.

I see that argument, but I think I disagree. I think we do need to change something. 

If a multi versioned functions signature is is dominated by the default (ie, that's the canonical signature) than I dont think it makes sense to allow calling that function without that signature in scope.

For example
```C
// TU 1
#import TU2

int fn (int c = 1);

int bar() { return fn(); }

// TU 2

int __attribute__((target_version("sve"))) fn (int c = 2);
int __attribute__((target_version("sve2"))) fn (int c = 2);

int bar() { return fn(); }
```
Would be valid and the calls would have different defaults.

I think the above is quite a good example of why the default declaration determining scope is a good idea.

I also don't think its unreasonable to require the other TU's to have hoisted/shared default declarations. That is the standard for C/C++ functions called in multiple files, and if foo calls bar, and bar calls foo, this would also be required no? It seems idiomatic to require it?

Reply:
I think I would be okay with a rule that only allows calling a multi-versioned function if you can see the default declaration: we can add diagnostics + fixits that help users do the right thing.